model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapStorageProviderFactory.java

```diff
@@ -156,14 +162,26 @@ public class JpaMapStorageProviderFactory implements
         MODEL_TO_TX.put(RoleModel.class,                        JpaRoleMapKeycloakTransaction::new);
     }
 
+    public JpaMapStorageProviderFactory() {
+        int index = ENUMERATOR.getAndIncrement();
+        this.sessionProviderKey = PROVIDER_ID + "-" + index;
+        this.sessionTxKey = SESSION_TX_PREFIX + index;
+    }
+
     public MapKeycloakTransaction createTransaction(Class<?> modelType, EntityManager em) {
         return MODEL_TO_TX.get(modelType).apply(em);
     }
 
     @Override
     public MapStorageProvider create(KeycloakSession session) {
         lazyInit();
-        return new JpaMapStorageProvider(this, session, emf.createEntityManager());
+        // check the session for a cached provider before creating a new one.
+        JpaMapStorageProvider provider = JpaMapStorageProvider.class.cast(session.getAttribute(this.sessionProviderKey));
+        if (provider == null) {
+            provider = new JpaMapStorageProvider(this, session, emf.createEntityManager(), this.sessionTxKey);
+            session.setAttribute(this.sessionProviderKey, provider);
+        }
```

Comment:
same here

Reply:
Yeah, I agree with the guard here as we can have concurrent calls to the same factory.create() method. Will add it.

Reply:
It is highly unlikely, you are right, usually there is only a single thread per session. Let's keep it as is.